File: src/main/java/upbrella/be/store/dto/response/StoreFindByIdResponse.java
package upbrella.be.store.dto.response;

import lombok.*;
import upbrella.be.store.entity.StoreDetail;

import java.time.LocalDateTime;
import java.util.List;
import java.util.stream.Collectors;

@Getter
@Builder
@NoArgsConstructor(access = AccessLevel.PRIVATE)
@AllArgsConstructor
public class StoreFindByIdResponse {

    private long id;
    private String name;
    private String category;
    private long availableUmbrellaCount;
    private boolean openStatus;
    private String businessHours;
    private String contactNumber;
    private String instaUrl;
    private String address;
    private String umbrellaLocation;
    private String description;
    private double latitude;
    private double longitude;
    private List<String> imageUrls;

    public static StoreFindByIdResponse fromStoreDetail(StoreDetail storeDetail, long availableUmbrellaCount) {

        return StoreFindByIdResponse.builder()
                .id(storeDetail.getId())
                .name(storeDetail.getStoreMeta().getName())
                .businessHours(storeDetail.getWorkingHour())
                .contactNumber(storeDetail.getContactInfo())
                .address(storeDetail.getAddress())
                .instaUrl(storeDetail.getInstaUrl())
                .description(storeDetail.getContent())
                .category(storeDetail.getStoreMeta().getCategory())
                .umbrellaLocation(storeDetail.getUmbrellaLocation())
                .availableUmbrellaCount(availableUmbrellaCount)
                .openStatus(storeDetail.getStoreMeta().isOpenStore(LocalDateTime.now())) // 의사 결정 전까지 임시로 True 고정
                .latitude(storeDetail.getStoreMeta().getLatitude())
                .longitude(storeDetail.getStoreMeta().getLongitude())
                .imageUrls(storeDetail.getSortedStoreImages().stream()
                        .map(storeImage -> storeImage.getImageUrl())
                        .collect(Collectors.toList()))
                .build();
    }
}

File: src/main/kotlin/upbrella/be/store/dto/response/StoreFindByIdResponse.kt
package upbrella.be.store.dto.response

import upbrella.be.store.entity.StoreDetail
import java.time.LocalDateTime

data class StoreFindByIdResponse(
    // TODO: 엔티티부터 nullable 필드 확인 필요
    val id: Long?,
    val name: String?,
    val category: String?,
    val availableUmbrellaCount: Long,
    val openStatus: Boolean?,
    val businessHours: String?,
    val contactNumber: String?,
    val instaUrl: String?,
    val address: String?,
    val umbrellaLocation: String?,
    val description: String?,
    val latitude: Double?,
    val longitude: Double?,
    val imageUrls: List<String>
) {
    companion object {
        fun fromStoreDetail(storeDetail: StoreDetail, availableUmbrellaCount: Long): StoreFindByIdResponse {
            return StoreFindByIdResponse(
                id = storeDetail.storeMeta?.id,
                name = storeDetail.storeMeta?.name,
                category = storeDetail.storeMeta?.category,
                availableUmbrellaCount = availableUmbrellaCount,
                openStatus = storeDetail.storeMeta?.isOpenStore(LocalDateTime.now()),
                businessHours = storeDetail.workingHour,
                contactNumber = storeDetail.contactInfo,
                instaUrl = storeDetail.instaUrl,
                address = storeDetail.address,
                umbrellaLocation = storeDetail.umbrellaLocation,
                description = storeDetail.content,
                latitude = storeDetail.storeMeta?.latitude,
                longitude = storeDetail.storeMeta?.longitude,
                imageUrls = storeDetail.storeImages.map {
                    requireNotNull(it.imageUrl) {
                        "Image URL should not be null"
                    }
                }
            )
        }
    }
}
